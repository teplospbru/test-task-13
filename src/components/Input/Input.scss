@import '../../scss/variables';
@import '../../scss/fonts';

.line-1 {
    @include flex(space-between, center);
    border-bottom: 1px solid $light-color;
    > div {
        margin: 0 31px;
        width: 48px;
        height: 48px;
        

        @media  screen and (max-width: 1024px) {
            margin: 0 21px;
            width: 32px;
            height: 32px;
        }
        svg {
            transition: 0.2s all;
            fill: $light-color;
        }
        &:hover svg {
            fill: $red-color;
        }
    }
    input {
        width: 100%;
        border: none;
        padding: 1.2rem 1em 1.2rem 0;
        outline: none;
        @include font(32, 300, 1.3);
        font-style: italic;
        color: $primary-color;
        &::placeholder {
            color: $light-color;
        }

        @media  screen and (max-width: 1024px) {
            @include font(24, 300, 1.3);
            font-style: italic;
        }
    }
}