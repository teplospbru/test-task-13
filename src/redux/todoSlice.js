import { createSlice } from "@reduxjs/toolkit";

export const todoSlice = createSlice({
    name: 'todos',
    initialState: {
        todos: [
            { id: 1, body: 'Next, we need to import the reducer function from the counter slice and add it to our store', isCompleted: false },
            { id: 2, body: 'Creating a slice requires a string name to identify the slice', isCompleted: true },
            { id: 3, body: 'Provide the Redux Store to React', isCompleted: false },
        ],
    },
    reducers: {
        createTodo(state, action) {
            state.todos.push({ id: Date.now(), body: action.payload, isCompleted: false })
        },
        setCompleted(state, action) {
            const index = state.todos.findIndex(todo => todo.id == action.payload);
            state.todos[index].isCompleted = !state.todos[index].isCompleted;
        },
        clearCompleted(state, action) {
            state.todos = state.todos.filter(todo => todo.isCompleted === false);
        }
    }
});

export const { clearCompleted, createTodo, setCompleted } = todoSlice.actions;

export default todoSlice.reducer;