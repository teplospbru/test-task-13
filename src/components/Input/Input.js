import './Input.scss';
import { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { createTodo } from '../../redux/todoSlice';

const Input = () => {
    const [ value, setValue ] = useState('');
    const [ isFocused, setFocused ] = useState(false);
    const [ isWarning, setWarning ] = useState(false);
    const dispatch = useDispatch();

    useEffect(() => {
        if(isWarning) {
            setTimeout(() => {
                setWarning(false);
            }, 2000);
        }
    }, [isWarning])

    const inputHandler = e => {
        setValue(e.target.value);
        if(e.target.value.length >= 3) {
            setWarning(false);
        }
    }

    const focusHandler = e => {
        setFocused(true);
    }

    const blurHandler = e => {
        setFocused(false);
    }

    const pressEnterHandler = e => {
        if(e.key === 'Enter' && isFocused && value.length < 3) {
            setWarning(true);
        }
        if(e.key === 'Enter' && isFocused && value.length >= 3) {
            setWarning(false);
            dispatch(createTodo(value));
            setValue('');
        }
    }

    const clickHandler = () => {
        if(value.length < 3) {
            setWarning(true);
        }
        if(value.length >= 3) {
            setWarning(false);
            dispatch(createTodo(value));
            setValue('');
        }
    }

    return (
        <div className='line-1'>
            {
                isWarning 
                    ? (
                        <div data-testid='warning'>
                            <svg>
                                <use xlinkHref="#warning"></use>
                            </svg>
                        </div> )
                    : (
                        <div style={{ cursor: 'pointer' }} onClick={ clickHandler } data-testid='btn'>
                            <svg>
                                <use xlinkHref="#chevron-1"></use>
                            </svg>
                        </div> )
            }
            <input 
                placeholder={ isWarning ? 'Enter 3 characters at least!' : 'What needs to be done?' } 
                value={ value }
                onChange={ inputHandler } 
                onKeyPress = { pressEnterHandler } 
                onFocus={ focusHandler }
                onBlur={ blurHandler }
            ></input>
        </div>
    )
};

export default Input;