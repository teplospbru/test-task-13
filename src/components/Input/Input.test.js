import Input from './Input';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';
import renderer from 'react-test-renderer';

beforeEach(() => {
    render(
        <Provider store={ store }>
            <Input />
        </Provider>
    )
})

describe('testing the component "Input"', () => {
    it('renders "Input"', () => {
        const input = screen.getByPlaceholderText('What needs to be done?');

        expect(input).toBeInTheDocument();
    });
    it('shows text in input field on change input value', () => {
        const input = screen.getByPlaceholderText('What needs to be done?');

        fireEvent.change(input, { target: { value: 'er' } });
        expect(screen.getByDisplayValue('er')).toBeInTheDocument();
    });
    ['', 'e', 'er'].forEach(item => {
        it(`shows alert on change input value to "${item}" and press button (must be 3 character at least to save todo)`, () => {
            const input = screen.getByPlaceholderText('What needs to be done?');
            const btn = screen.getByTestId('btn')
    
            fireEvent.change(input, { target: { value: item } });
            fireEvent.click(btn);

            expect(screen.getByTestId('warning')).toBeInTheDocument();
            expect(screen.getByPlaceholderText('Enter 3 characters at least!')).toBeInTheDocument();
        });
    });
    ['', 'e', 'er'].forEach(item => {
        it(`shows alert on change input value to "${item}" and press "Enter" (must be 3 character at least to save todo)`, () => {
            const input = screen.getByPlaceholderText('What needs to be done?');
    
            fireEvent.change(input, { target: { value: item } });
            fireEvent.focus(input);
            fireEvent.keyPress(input, { key: 'Enter', charCode: 13 });

            expect(screen.getByTestId('warning')).toBeInTheDocument();
            expect(screen.getByPlaceholderText('Enter 3 characters at least!')).toBeInTheDocument();
        });
    });
    it('removes alert while alert is showed and input value changed from "er" to "ers"', () => {
        const input = screen.getByPlaceholderText('What needs to be done?');

        fireEvent.change(input, { target: { value: 'er' } });
        fireEvent.click(screen.getByTestId('btn'));

        expect(screen.getByTestId('warning')).toBeInTheDocument();

        fireEvent.change(input, { target: { value: 'ers' } });

        expect(screen.getByTestId('btn')).toBeInTheDocument();
    });
    it('matches to snapshot', () => {
        const container = renderer.create(<Provider store={ store }><Input /></Provider>).toJSON();

        expect(container).toMatchSnapshot()
    })
})