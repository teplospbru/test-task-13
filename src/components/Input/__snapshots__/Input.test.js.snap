// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the component "Input" matches to snapshot 1`] = `
<div
  className="line-1"
>
  <div
    data-testid="btn"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <svg>
      <use
        xlinkHref="#chevron-1"
      />
    </svg>
  </div>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="What needs to be done?"
    value=""
  />
</div>
`;

exports[`testing the component "Input" renders the component 1`] = `
<div
  className="line-1"
>
  <div
    data-testid="btn"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <svg>
      <use
        xlinkHref="#chevron-1"
      />
    </svg>
  </div>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="What needs to be done?"
    value=""
  />
</div>
`;
